Build pipeline refactor: Image, Local and VM strategies

namesty opened the pull request:
The current PR closes #1246 and #1119.

# Overview

It introduces a refactor to the build pipeline composed of 3 main changes:

## Strategies

Implementation of different build "strategies", which are abstractions of different source building algorithms. This also implied a decoupling of Docker from the `Compiler` and `Project`. Currently there are 3, and "VM" is the new default:

- `Image`: default approach, prior to this PR. It implies building a Docker image at runtime, where dependencies are installed and sources are copied and built as `Dockerfile` instructions. On subsequent builds, Docker tries to reuse cached image layers and rebuild accordingly. This approach is notably slow but the complete process happens in Docker, and can be reproduced, examined and audited layer by layer (from dependency installation to build artifacts output)

- `Local`: approach that does not use Docker at all. It simply executes a `.sh` file that contains the necessary instructions to install dependencies and build sources. While this is the fastest way of building, it involves requiring the user to install "prequisite" system dependencies and given that sources are built on the host machine and not a reproducible docker environment, "reproducibility" isn't guaranteed ("but it builds on my machine tho!"). This strategy does not concern itself with dependency symbolic links, as all dependencies are handled by the host's dep manager.

- `VM`: approach that uses Docker only for the source building part of the process. At build time, it pulls a pre-built image with all necessary system dependencies, env vars and runtime; and it instantiates a Docker container with it. The Docker container instantiates bind-mounts (volumes) to copy the sources and dependencies from the host, build the sources inside the container, and copy the build artifacts back to the host machine. This approach ensures that the sources will be built in a reproducible environment but it doesn't use Docker for anything else and no image is built at runtime.

## Build Manifest Changes

The build manifest was changed to contemplate configurations for build strategies, and strategy-specific overrides/config. It also validates strategy configurations based on the Polywrap's manifest language. Example: if the project is an Assemblyscript wrapper and the image strategy is going to be used, then `node_version` is a required image-strategy config property; but it is not if the project is a Rust wrapper.

## Different handling of build cache (`.polywrap/build`) directory

Now the defaults (scripts, templates, etc.) used for any strategy are cached in the `.polywrap/build` dir for easy debugging and overriding. We no longer write a default manifest to the cache folder, only to read it later with its manifest extension.

Other misc impl details include: `Compiler` does not generate code but depends on `CodeGenerator`, compiler overrides.

# Short-term future needs

This PR introduces some short-term needs we need to cover in other PRs:

- Build benchmarks repository and automated measurements and stress testing (so far, I've been just locally testing)
- Automated pipeline to change and re-publish base VM images to DockerHub

# Further fleshing-out of VM strategy

https://hackmd.io/HQHG0xXfT4CEc-iHsuMHpg

namesty committed:
(feat): Source build strategies. Extracted Docker part from compiler

namesty committed:
(wip): removed "CompilerState" and possible related errors

namesty committed:
(chore): schemaComposer removed from CodeGenerator deps

namesty committed:
(fix): fixed sourceBuildStrategy passing to getCompilerOverrides

namesty committed:
(chore): updated strategies. Abstracted more docker related funcs

namesty committed:
(feat): added no-docker build option. Lock and overrides

namesty committed:
(chore): removed CompilerOverrides concept. All codegen to generator

namesty committed:
(chore): unified compiler/generator code generation. Tests passing

namesty committed:
(feat): codegen strategies

namesty committed:
(chore): updated help text for build in tests

namesty committed:
(chore): made schemaComposer a dep again. All tests passing

namesty committed:
(chore): fixed negatable option flag

namesty committed:
(chore): copying build scripts to build folder

nerfZael commented on packages/cli/src/commands/build.ts:
```diff
@@ -67,43 +70,76 @@ export const build: Command = {
   const execute = async (): Promise<boolean> => {
-    compiler.reset();
+    project.reset();
+    schemaComposer.reset();
```
Can't say for sure without diving deeper, but these resets feel sketchy

namesty replied:
They were, and were serving no purpose; just removed them

nerfZael commented on packages/cli/src/lib/codegen/strategies/CompilerCodegenStrategy.ts:
```diff
@@ -0,0 +1,37 @@
+    return writeDirectorySync(binding.outputDirAbs, binding.output);
+  }
+
+  private async _getGenerationSubpath() {
```
Missing return type

namesty replied:
Woo, nice catch

nerfZael commented on packages/cli/src/commands/build.ts:
```diff
@@ -67,43 +70,76 @@ export const build: Command = {
+  await validateManifestModules(polywrapManifest);
 
-  const schemaComposer = new SchemaComposer({
+  const buildStrategy = createSourceBuildStrategy(docker ? "docker" : "local", {
```
It might even be clearer if there was no `createSourceBuildStrategy` func and it's code was just here.
Wdyt?

namesty replied:
Yep, agreed

pileks commented on packages/cli/src/commands/codegen.ts:
```diff
@@ -82,25 +82,23 @@ async function run(options: CodegenCommandOptions) {
+        omitHeader: false,
+        mustacheView: undefined,
+      })
+    : new CompilerCodegenStrategy({
```
This is the line I was most interested in.

In my PR, it's all reduced to the following:
https://github.com/polywrap/toolchain/pull/1051/files#diff-87c9d295ac1ee17a0b50c70f922ce1ebc7ad5da4746b6d4f18b45334452ae030R106

It seems to work... 😁

namesty replied:
Yep, you're right. In fact, between your PR and this one; I think we're removing significant dead code

namesty committed:
(chore): updated build scripts

namesty committed:
(chore): updated local strategy

namesty committed:
(chore): eliminated createSourceBuildStrategy

namesty committed:
(chore): intl messages for local build strategy

namesty committed:
(chore): moved docker build related functions from utils to strategy

namesty committed:
(fix): rust script env variables and AS shebang

namesty committed:
(chore): runCommand CWD override

namesty committed:
(feat): added basic test cases for no docker

namesty committed:
(chore): improved naming and messaging

namesty committed:
(chore): added sanity cases for all strategies in all languages

namesty committed:
(chore): removed resets

namesty committed:
Merge remote-tracking branch 'origin/origin-dev' into namesty/docker-build-vm

namesty committed:
(chore): removed compiler codegen strategy

namesty committed:
(chore): removed codegen strategies in favor of normal inheritance

namesty committed:
(chore): added Rust sanity test cases for codegen

namesty committed:
(chore): minor fix to build test cases dir structure on sanity cases

namesty commented:
After seeing @pileks's changes, indeed the Compiler's codegen does not anything really different to the CodeGenerator's codegen. Additionally, the `generationSubpath` argument may, at some point, be used but right now it is not being used since we're overriding the default `src/wrap` with `src/wrap` if we're codegenerating for Rust. For these reasons, (and the lack of need for changing strategies at runtime) I just removed the strat pattern for codegen, in favor of simpler, classic inheritance.

I noticed we did not have sanity tests for Rust, when using the `build` command or the `codegen` command; so to really make sure this refactor does not cause regressions, I added sanity cases for Rust for both builds and codegens.

I decided to keep this PR smaller by just incluiding the "no docker" build possibility; I'll push the Docker VM approach in a different PR.

namesty committed:
(fix): added wasm-snip implicit install

namesty committed:
(fix): script code generation now gets abi

namesty committed:
(fix): docker lock

namesty committed:
(fix): removed unused stdout

namesty committed:
(chore): spinner for local sources build

namesty commented:
Poured my performance analysis on this PR here: https://hackmd.io/IVTtH4rTSa6ud5XwIUkCDw

namesty committed:
(temp): CI test msg logging

namesty committed:
(temp): CI CLI speed up

namesty committed:
(chore): updated test cases

namesty committed:
(fix): fix relative path to linked runtime packages

namesty committed:
(fix): minor fix to test ignore patterns

namesty committed:
(feat): CLI rust tests GH action

namesty committed:
(chore): removed unused dependencies

namesty committed:
(wip): scripts and images. Initial logic

namesty committed:
(chore): naive strategy for assemblyscript

namesty committed:
(fix): wasm-rs relative path. Build only if module. Removed unncessary arg

namesty committed:
Merge branch 'namesty/docker-build-vm' into namesty/docker-vm

namesty committed:
(chore): Language specific helpers for script actions

namesty committed:
(chore): removed scriptHelpers in favor of script files

namesty committed:
Merge branch 'namesty/docker-vm-scripts' into namesty/docker-vm

namesty committed:
Merge remote-tracking branch 'origin/origin-dev' into namesty/docker-vm

namesty committed:
(chore): solved merge conflicts. Unused codegen method

namesty committed:
(chore): project being built and artifacts being copied

namesty committed:
(chore): added spinners, copy artifacts messaging. Test passing

namesty committed:
(chore): added rust build script template

namesty commented:
Closes #1119

namesty committed:
(chore): updated Rust strategy + Updated Docker base image for rust

namesty committed:
(chore): added tests for strategy in Rust and AS

namesty committed:
Merge branch 'origin-dev' into namesty/docker-vm

namesty committed:
Merge branch 'namesty/docker-vm' into namesty/docker-build-vm

namesty committed:
(chore): solved merged conflicts. Building

namesty committed:
(chore): changed generated dirs and files permissions

namesty committed:
(chore): added strategy overrides to build manifest

namesty committed:
(fix): new build manifest props

namesty committed:
(chore): migrated build manifests to 0.2.0

namesty committed:
(fix): jest rs config for cli fix

namesty committed:
(chore): no codegen test case added

namesty committed:
(chore): DockerStrategy -> ImageStrategy. Made VM the default strategy

namesty committed:
(fix): only copy additional includes if they exist

namesty committed:
(fix): default strategy help text. Filelock for image strategy only

namesty committed:
(chore): added default includes as overridable manifest config prop

namesty committed:
(chore): reorganized build defaults

namesty committed:
(chore): fetching vm configs from defaults dir

namesty committed:
(chore): once again changed manifest. Removed language keys

namesty committed:
(chore): decoupled docker specifics from PolywrapProject completely

namesty committed:
(fix): docker -> image

namesty committed:
(chore): improved tests by adding strategy.json. Img suffix check

namesty committed:
(chore): dynamic manifest validation per strategy based on language

namesty committed:
(chore): dep linking for AS projects

namesty committed:
(chore): updated tests. Added validation test for strategy+lang props

namesty committed:
(chore): using cmd.json

namesty committed:
(chore): better error handling for VM strategy

namesty committed:
(chore): reorganized defaults for build-strategies

namesty committed:
(chore): added used strategies to cache folder

namesty committed:
Merge branch 'origin-dev' into namesty/docker-build-vm

namesty committed:
(chore): removed unnecessary OR object

namesty committed:
(chore): updated all build manifests from 0.1.0 -> 0.2.0

namesty committed:
(temp): console.logs on rs-ci

pileks commented on packages/cli/src/lib/build-strategies/strategies/DockerVMStrategy.ts:
```diff
@@ -0,0 +1,254 @@
+  },
+};
+
+interface BuildManifestConfig {
```
This seems like a duplicate of [PolywrapProject.ts:42](https://github.com/polywrap/toolchain/pull/1204/files#diff-77f66692da7e2e7f64d7d60655863cba49e4b2fa2071510ae8d1a0e655266510R42)

Any specific reason for the duplication?

namesty replied:
Nope. Great catch on a merge remnant hehe

namesty committed:
(chore): skipped local strategy test

namesty committed:
Merge branch 'origin-dev' into namesty/docker-build-vm

namesty committed:
(chore): solve merge conflicts

namesty committed:
(chore): docker lock test fix

pileks commented on packages/cli/src/lib/build-strategies/strategies/ImageStrategy.ts:
```diff
@@ -0,0 +1,330 @@
+        if (cache == true) {
+          cacheDir = this.project.getCachePath("build/image/cache");
+        } else if (cache) {
+          if (!path.isAbsolute(cache)) {
```
Just from a style perspective, maybe omit the invert the if so that we don't have to read the `!` which might be hard to notice at a cursory glance? Nitpicking here...

namesty replied:
Yep, I agree with this

pileks reviewed (changes requested):
All in all looks great! See comments and we're good!

cbrzn commented on packages/cli/jest.rs.config.js:
```diff
@@ -0,0 +1,24 @@
+      diagnostics: false
+    },
+  },
+  modulePathIgnorePatterns: [
```
We have two definitions of `modulePathIgnores` and `testMatch`- I guess this should change right?

cbrzn commented on packages/cli/src/__tests__/e2e/build-rs.spec.ts:
```diff
@@ -0,0 +1,68 @@
+import fs from "fs";
+import path from "path";
+
+jest.setTimeout(500000);
```
It looks like this is not enough for CI when it comes to building w/docker :(

namesty replied:
I just double checked, jsut to make sure there was no sneaky deadlock; and indeed, it just takes an awful lot of time. Pumped the timeout by 2 mins

cbrzn commented on packages/cli/src/__tests__/e2e/build-rs.spec.ts:
```diff
@@ -0,0 +1,68 @@
+    });
+  })
+
+  // NOTE: Skipped because CI needs system prequisites: bindgen, snip, etc.
```
ignorant q: Would it be possible to install this in the CI script?

cbrzn replied:
now that I am thinking more about this, I think it would make sense to add instructions somehow to download the necessary dependencies if you'd like to run this build locally; if I am not wrong it would mean to run the `local.sh` from rust?

I guess you get the point - lmk if this doesn't make sense and i can try harder to explain myself

namesty replied:
Agreed. Yeah I was thinking about including the dependencies in a Readme or an install.sh script; because it shouldn't be the Polywrap's CLI responsibility to install these permanent system dependencies on the user's machine.

namesty replied:
I almost feel like, we could get this merged and open an issue on where to include that in the docs, or maybe add a command flag to output an install script for the user to run. Thoughts? @dOrgJelli

cbrzn replied:
> we could get this merged and open an issue on where to include that in the docs

yah lets do it

cbrzn commented on packages/cli/src/__tests__/e2e/build.spec.ts:
```diff
@@ -1,4 +1,3 @@
-import { PolywrapProject, loadBuildManifest } from "../../lib";
```
nit: I am noticing that now we have `build` and `build-rs`. Would it make sense to create a folder `build` inside of `__tests__` and inside of this add an implementation test?

My main thought here is that for example if we would like to add tests for `go` build, it seems better to just add `go.spec.ts` inside of the `__tests__/build` folder, rather than creating a `build-go.spec.ts`?

namesty replied:
I agree, but I think we should do this consistently for the other test suites as well, probably in another PR

cbrzn commented on packages/cli/src/commands/build.ts:
```diff
@@ -8,19 +8,24 @@ import {
+import { LocalBuildStrategy } from "../lib/build-strategies/strategies/LocalStrategy";
+import { ImageBuildStrategy } from "../lib/build-strategies/strategies/ImageStrategy";
+import { BuildStrategy, SUPPORTED_STRATEGIES } from "../lib/build-strategies";
+import { DockerVMBuildStrategy } from "../lib/build-strategies/strategies/DockerVMStrategy";
```
nit: add these guys to `lib/build-strategies/index.ts` and just have one `import` statement here? 

feel free to ignore this it's a style thing

cbrzn commented on packages/cli/src/commands/build.ts:
```diff
@@ -8,19 +8,24 @@ import {
+import { PolywrapManifest } from "@polywrap/polywrap-manifest-types-js";
 
 const defaultOutputDir = "./build";
+const defaultStrategy = "vm";
```
nit: what about converting this (& [supported strategies values](https://github.com/polywrap/toolchain/blob/namesty/docker-build-vm/packages/cli/src/lib/build-strategies/index.ts#L3)) to enum, so if you change the enum, you don't need to change it in any other file? :thinking:

cbrzn commented on packages/cli/src/commands/build.ts:
```diff
@@ -67,56 +79,91 @@ export const build: Command = {
           ),
           clientConfig: await parseClientConfigOption(options.clientConfig),
           outputDir: parseDirOption(options.outputDir, defaultOutputDir),
+          strategy: options.strategy ?? defaultStrategy,
```
nit: I think this can be removed since we have the `default` attribute in strategy option?

cbrzn commented on packages/cli/src/lib/build-strategies/BuildStrategy.ts:
```diff
@@ -0,0 +1,56 @@
+
+    if (
+      fse.existsSync(
+        this.project.getCachePath(PolywrapProject.cacheLayout.buildStrategyUsed)
```
real nit feel free to ignore:

what about converting `this.project.getCachePath....` into a variable and use this variables in all the cases, hence, making this code easier to read?

cbrzn commented on packages/cli/src/lib/codegen/CodeGenerator.ts:
```diff
@@ -0,0 +1,107 @@
+  }
+
+  // eslint-disable-next-line @typescript-eslint/naming-convention
+  protected async runCodegen(_: BindLanguage, __?: Ora): Promise<string[]> {
```
Probably I am missing something and am wrong here but do we need these arguments in this function?

namesty replied:
The base class may not use them, but they're still available for classes that inherit from this one. For example: `ScriptCodegen` uses them

cbrzn commented on packages/cli/src/lib/codegen/ScriptCodeGenerator.ts:
```diff
@@ -0,0 +1,120 @@
+  private _mustacheView: Record<string, unknown> | undefined;
+  private _codegenDirAbs: string;
+  private _omitHeader: boolean;
+  private _schema: string | undefined = "";
```
nit: all these attributes can be `readonly`

cbrzn commented on packages/cli/src/lib/project/PolywrapProject.ts:
```diff
@@ -41,12 +39,28 @@ export interface PolywrapProjectConfig extends ProjectConfig {
   metaManifestPath?: string;
 }
 
+export interface BuildManifestConfig {
```
Same q: https://github.com/polywrap/toolchain/pull/1204#discussion_r986323786

cbrzn commented on packages/cli/src/lib/project/manifests/polywrap/languages.ts:
```diff
@@ -37,3 +38,21 @@ export function polywrapManifestLanguageToBindLanguage(
   }
 }
+
+export async function getGenerationSubpath(
```
Looks like this function's not being used

cbrzn commented on packages/js/plugins/ethereum/src/__tests__/integration/polywrap.build.yaml:
```diff
@@ -1,6 +1,9 @@
+format: 0.2.0
+strategies:
+  image:
+    node_version: "14.16.0"
```
nit: would it be possible to use `16.13.0` in all the custom build manifests (for as) in monorepo?

namesty replied:
I think this should probably be done in a separate PR

cbrzn commented on packages/js/manifests/polywrap/src/formats/polywrap.build/0.2.0.ts:
```diff
@@ -0,0 +1,101 @@
+  /**
+   * Polywrap build manifest format version.
+   */
+  format: "0.2.0" | "0.2";
```
i think we should not support `0.2`

cbrzn commented on packages/test-cases/cases/cli/wasm/build-cmd/assemblyscript/001-sanity/yarn.lock:
```diff
@@ -0,0 +1,51 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
Remove

cbrzn reviewed (commented):
:fire: regarding next steps i can def help with the benchmark tool :)

cbrzn commented:
btw just noticed you're fixing 19 alerts (https://github.com/polywrap/toolchain/pull/1204#issuecomment-1267123319) :goat:

namesty committed:
(chore): removed yarn lock from AS test case

namesty committed:
(chore): removed 0.2 support. Removed unused functions

namesty committed:
(chore): removed duplicate interface

namesty committed:
(chore): refactored imports. Inverted not-condition

namesty committed:
(chore): readonly props for codegenerator

namesty committed:
(chore): changed string literals array to enum

namesty committed:
(chore): reused variable, instead of using getCache everytime

namesty committed:
(chore): removed unnecessary/duplicated props from CLI Rs jest config

namesty committed:
(chore): commander default strategy opt

namesty committed:
(test): Docker CI test temp

namesty committed:
(chore): uncommented skips
